Review report draft

MarekSuchanek opened the pull request:
Thank you all for your contributions! Now is the time for final review before submission.

Please go through the report as a whole. You can either review the file directly in [Markdown](https://github.com/ds-wizard/bh2021-report/pull/5/files#diff-0403ef06adf405f7b310b4518bd6a3559854f54c61676f676ce9cbfee7172ab6) or [download the PDF](https://github.com/ds-wizard/bh2021-report/suites/4650718283/artifacts/126219499). Please submit the review **before 14 Dec**. I will incorporate required changes and submit the report to BioHackrXiv.

❗ **Do not push now to master**, use just GitHub review and/or change suggestions (see [docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/about-pull-request-reviews) if needed), so we all review the same version:

- ✔️  **Approve** = you are approving the report to be submitted as is.
- 💬  **Comment** = you need to clarify something but it looks fine in general for you.
- ❌ **Request changes** = you need the report to have some changes (that you clearly explain or suggest) before the submission.

Use discussion in PR below if needed as well 🙂 

💡 Also check if your name, ORCID, and affiliation is correct and according to your with.

Commit:
starting on dpia knowlegde model

Commit:
KM report first draft

Commit:
fix typos

MarekSuchanek committed:
Add old integration and doc HTTP

MarekSuchanek committed:
Add original integration and HTTP jinja figures

janslifka committed:
Add widget description

Commit:
motivations with DSW

MarekSuchanek committed:
Add integration design and conclusions

MarekSuchanek committed:
Add authors short

Commit:
Slight rewrite of Abstract

Commit:
Slight rewrite of Introduction, part 1.

Commit:
Slight rewrite of Introduction, part 2.

Commit:
Slight rewrite of DPIA in DSW

Commit:
Slight rewrite of DPIA KM, part 1.

Commit:
Slight rewrite of DPIA KM, part 2.

Commit:
Slight rewrite of DPIA KM, part 2; more...

Commit:
Slight rewrite of output template

Commit:
Slight rewrite of DPIA projects

vildead committed:
extend of 'DPIA in DSW' section

vildead committed:
small rephrase in DSW Widget Integrations

vildead committed:
Widget for DAISY Projects section

vildead committed:
add link to CNIL PIA tool

vildead committed:
update of section 'Creating a DPIA Knowledge Model

- add comment on risk assessment
- minor adjustments

Commit:
dpia document template

vildead committed:
Merge pull request #3 from ds-wizard/widget-for-daisy-projects

Widget for daisy projects

vildead committed:
minor updates Creating a DPIA Knowledge Model section

vildead committed:
major update of DPIA Document Template

- sentences describing general DSW template description are removed
- reason for selection of CNIL templated is explained

vildead committed:
remove section DPIA Projects in DSW

- general description of how KM-template works in DSW is imho not needed
- if there is anything specific for our KM-template interaction, it can
be included in previous section

vildead committed:
addressing review comments from Paulette

vildead committed:
capitalizing 'Knowledge Model'

vildead committed:
adding links to DPIA template

MarekSuchanek committed:
Merge pull request #4 from ds-wizard/dpia_knowledge_model

DPIA knowledge model

MarekSuchanek committed:
Add ORCID for Marko and Nadja

MarekSuchanek committed:
Minor consistency improvements

froggypaule commented on paper.md:
```diff
@@ -38,89 +38,140 @@ affiliations:
 # DPIA in DSW
 
-*TODO: (Pinar) and Vilém - motivation for having DPIA KM+template in DSW, core idea and advantages*
+Writing a data protection impact assessment for a scientific project is a challenging task. Input from multiple key groups is required to obtain complete and precise information. In addition to data stewards, which are commonly in charge of the information collection since their expertise overlaps all other concerned domains, the process involves researchers themselves bringing their own knowledg ...
```
the process involves researchers themselves bringing their own knowledge about the data under consideration and the related scientific processes.

MarekSuchanek replied:
```suggestion
Writing a data protection impact assessment for a scientific project is a challenging task. Input from multiple key groups is required to obtain complete and precise information. In addition to data stewards, which are commonly in charge of the information collection since their expertise overlaps all other concerned domains, the process involves researchers themselves bringing their own knowledge about the data under consideration and the related scientific processes. Data managers and IT staff who are knowledgeable about the local IT infrastructure, and external collaborators may also report on the shared data processing tasks. Finally, legal representatives (e.g. data protection officers) validate the resulting document using existing contracts and institutional policies.
```

froggypaule commented on paper.md:
```diff
@@ -38,89 +38,140 @@ affiliations:
+* The measures taken to protect the rights of the subjects
+* The identification, analysis, and evaluation of the data risks factors and the measures and controls to mitigate those risks
+
+Although a risk assessment is a very important part of the DPIA, our Knowledge Model is addressing it very briefly by questions on general controls which are or will be in place. As risks generally depend on the infrastructure used and the institutional units involved, they would apply globally to all the processing under their authority. We propose a process where the risk assessment matrix can  ...
```
our Knowledge Model is addressing it only very briefly

MarekSuchanek replied:
```suggestion
Although a risk assessment is a very important part of the DPIA, our Knowledge Model is addressing it only very briefly by questions on general controls which are or will be in place. As risks generally depend on the infrastructure used and the institutional units involved, they would apply globally to all the processing under their authority. We propose a process where the risk assessment matrix can be generated by in-house or external professionals using dedicated tools (e.g. [Monarc](https://www.monarc.lu/)) and reused for each DPIA as a resource. For this purpose, our Knowledge Model includes a question to provide link to such a document.
```

froggypaule commented on paper.md:
The other independent approach that we decided to implement uses linking a DAISY project from DSW and then querying data using DAISY API when needed. The overal idea is shown in Figure 1.
A use case may be described as follows. A user of DSW works on a project (with a goal to get a DMP but not limited to it) and in the questionnaire selects a question that uses an API integration to a project in DAISY.  That is, answering to this question will trigger an API call to retrieve project data from DAISY. At the moment of document production, that is, producing the actual DMP or DPIA report, current DAISY project data is queried using a DAISY REST API and is formatted using the desired template. The user then gets a document with data from the DAISY project appropriately included within other data originating directly from answering questions in the DSW project.

MarekSuchanek replied:
```suggestion
The other independent approach that we decided to implement uses linking a DAISY project from DSW and then querying data using DAISY API when needed. The overal idea is shown in Figure 1. A use case may be described as follows. A user of DSW works on a project (with a goal to get a DMP but not limited to it) and in the questionnaire selects a question that uses an API integration to a project in DAISY. That is, answering to this question will trigger an API call to retrieve project data from DAISY. At the moment of document production, that is, producing the actual DMP or DPIA report, current DAISY project data is queried using a DAISY REST API and is formatted using the desired template. The user then gets a document with data from the DAISY project appropriately included within other data originating directly from answering questions in the DSW project.
```

froggypaule commented on paper.md:
With such an approach, data in DAISY can be independently changed without modifying the DSW project: only at the time of document creation is the updated DAISY data queried and included in the document. Thus, a document creation is a snapshot of both DSW and DAISY projects at the time of the creation. 
To implement this feature, we first needed to investigate how the DSW API call functionality, also called 'integration question', can be put into practice and extended to allow project selection in DAISY.  We then had to solve how to format the DAISY data with a DSW document template that uses [Jinja2 language](https://jinja.palletsprojects.com).

froggypaule replied:
This reminds me: it is the first time I understand this. Question for the future: could we envisage DAISY data populating DMP project (ie before document creation)?

MarekSuchanek replied:
```suggestion
With such an approach, data in DAISY can be independently changed without modifying the DSW project: only at the time of document creation is the updated DAISY data queried and included in the document. Thus, a document creation is a snapshot of both DSW and DAISY projects at the time of the creation.  To implement this feature, we first needed to investigate how the DSW API call functionality, also called 'integration question', can be put into practice and extended to allow project selection in DAISY.  We then had to solve how to format the DAISY data with a DSW document template that uses [Jinja2 language](https://jinja.palletsprojects.com).
```

froggypaule commented on paper.md:
DSW supports integration questions since v1.6.0 (April 2019). It allows to query a REST API for type hints – as users type a reply, answers are suggested by the API response from which one specific answer can be chosen. The advantage is configurability of the integrations and the stored ID of selected item. On the other hand, it does not allow any form of user authentication towards the external API nor more complex item selection than just a regular input dropdown. In Figure 2, you can see how the integration works for selecting affiliation from [Research Organization Registry (ROR)](https://ror.org).



"The advantage is configurability of the integrations and the stored ID of selected item.": I don't understand what is meant there?

MarekSuchanek replied:
```suggestion
DSW supports integration questions since v1.6.0 (April 2019). It allows to query a REST API for type hints – as users type a reply, answers are suggested by the API response from which one specific answer can be chosen. The advantage is configurability of the integrations and the stored ID of selected item. On the other hand, it does not allow any form of user authentication towards the external API nor more complex item selection than just a regular input dropdown. In Figure 2, you can see how the integration works for selecting affiliation from [Research Organization Registry (ROR)](https://ror.org).
```

froggypaule commented on paper.md:
We attempted to use this integration feature as is to query DAISY projects. However, it has two pitfalls:

MarekSuchanek replied:
```suggestion
We attempted to use this integration feature as is to query DAISY projects. However, it has two pitfalls:
```

froggypaule commented on paper.md:
To access the API, a secret token must be used for authentication. Everyone wanting to include an API call into a such a Knowledge Model question would need to acquire such a token and place it into the DSW server configuration file.

MarekSuchanek replied:
```suggestion
1. To access the API, a secret token must be used for authentication. Everyone wanting to include an API call into a such a Knowledge Model question would need to acquire such a token and place it into the DSW server configuration file.
```

froggypaule commented on paper.md:
what is the relationship between 1 & 2: that one cannot implement authentication?
if that is the case, finish the previous paragraph by:
This is obviously not practical and one must thus overcome the need for authentication.

froggypaule replied:
As there is no way to distinguish specific users, the call response will consist of all (published) projects from DAISY. The natural behavior -- displaying only projects related to the user filling the questionnaire -- is not achievable.

"behavior" or "behaviour": do you use US English or UK English? importantly, one must stay consistent.
From my limited expertise, in France one is supposed to use UK English (as in Australia for example)

MarekSuchanek replied:
```suggestion
2. As there is no way to distinguish specific users, the call response will consist of all (published) projects from DAISY. The natural behaviour -- displaying only projects related to the user filling the questionnaire -- is not achievable.
```

I will check UK vs US before submission (will use UK English).

froggypaule commented on paper.md:
Therefore, we came up with the idea of an integration widget. Instead of a search field in DSW, we open a popup window with a web page implemented on the connected service side, in this case, DAISY. The implementation of the widget itself is the responsibility of the connected service. This means that any necessary functionality, such as login, can be implemented.

froggypaule replied:
bravo! again, it is the first time I understand what you have being doing

MarekSuchanek replied:
```suggestion
Therefore, we came up with the idea of an integration widget. Instead of a search field in DSW, we open a popup window with a web page implemented on the connected service side, in this case, DAISY. The implementation of the widget itself is the responsibility of the connected service. This means that any necessary functionality, such as login, can be implemented.
```

froggypaule commented on paper.md:
The widget implemented on DAISY's side allows users to select one of the existing projects and send them back as an answer to DSW. When the widget is opened in a popup window, users are asked to log into DAISY. As the initial page is the actual DAISY login form, the DAISY login credentials are required. Once logged in, users are redirected to the `/integrations/dsw/list-projects` DAISY endpoint presenting a list of projects they have access to. The list contains only records for which users are appointed to as local custodians or personnel. The internal DAISY project ID, project's acronym and full title is presented. When a project of interest is clicked on, the popup window is closed, and the information is sent to DSW.

MarekSuchanek replied:
```suggestion
The widget implemented on DAISY's side allows users to select one of the existing projects and send them back as an answer to DSW. When the widget is opened in a popup window, users are asked to log into DAISY. As the initial page is the actual DAISY login form, the DAISY login credentials are required. Once logged in, users are redirected to the `/integrations/dsw/list-projects` DAISY endpoint presenting a list of projects they have access to. The list contains only records for which users are appointed to as local custodians or personnel. The internal DAISY project ID, project's acronym and full title is presented. When a project of interest is clicked on, the popup window is closed, and the information is sent to DSW.
```

froggypaule commented on paper.md:
To allow for retrieving of project data from DAISY using its REST API, we had to allow making HTTP requests from Jinja2 templates. This is now a new feature of the [DSW Document Worker](https://github.com/ds-wizard/document-worker) component. As it is written in [Python](https://www.python.org) programming language, we decided to use the well-known [Requests library](https://docs.python-requests.org/en/latest/). Because any data steward can upload their own template and creating requests might be harmful, we also designed a configuration for the document worker that can enable this feature for specific templates (based on their IDs), set limit and timeout for the HTTP requests made from Jinja2.


again, because I surely don't get everything here: could this not be implemented from the KM/project itself, instead of the template?

MarekSuchanek replied:
```suggestion
To allow for retrieving of project data from DAISY using its REST API, we had to allow making HTTP requests from Jinja2 templates. This is now a new feature of the [DSW Document Worker](https://github.com/ds-wizard/document-worker) component. As it is written in [Python](https://www.python.org) programming language, we decided to use the well-known [Requests library](https://docs.python-requests.org/en/latest/). Because any data steward can upload their own template and creating requests might be harmful, we also designed a configuration for the document worker that can enable this feature for specific templates (based on their IDs), set limit and timeout for the HTTP requests made from Jinja2.
```

froggypaule commented on paper.md:
To make limits and timeouts effective, we had to introduce a simple wrapper class around the requests library. An object of such a class is then passed as a variable into a Jinja2 template if the feature is enabled. On each request through the object, a counter is increased and checked against the limit. The configured timeout is passed to the call of Requests library functions. It does not support sessions as it is expected that just a single request (or few consecutive requests) will be made to optimise document generation speed. A HTTP request can be made from a Jinja2 template in the way shown in Figure 5.

MarekSuchanek replied:
```suggestion
To make limits and timeouts effective, we had to introduce a simple wrapper class around the requests library. An object of such a class is then passed as a variable into a Jinja2 template if the feature is enabled. On each request through the object, a counter is increased and checked against the limit. The configured timeout is passed to the call of Requests library functions. It does not support sessions as it is expected that just a single request (or few consecutive requests) will be made to optimise document generation speed. A HTTP request can be made from a Jinja2 template in the way shown in Figure 5.
```

froggypaule commented on paper.md:
Using the HTTP requests enabled in document templates, we implemented the [DPIA Appendix Template](https://github.com/ds-wizard/daisy-appendix-template) that can be easily re-used in other DSW document templates. It is based on the example Knowledge Model where there is only a single question that asks for the DAISY project using the widget integration. However, the template is designed in a way that its central part is a re-usable Jinja2 `macro`. Anyone can take this part of the template, include it in their own one and call this macro using the DAISY project identifier as an argument.

MarekSuchanek replied:
```suggestion
Using the HTTP requests enabled in document templates, we implemented the [DPIA Appendix Template](https://github.com/ds-wizard/daisy-appendix-template) that can be easily re-used in other DSW document templates. It is based on the example Knowledge Model where there is only a single question that asks for the DAISY project using the widget integration. However, the template is designed in a way that its central part is a re-usable Jinja2 `macro`. Anyone can take this part of the template, include it in their own one and call this macro using the DAISY project identifier as an argument.
```

froggypaule commented on paper.md:
The macro itself makes three HTTP requests to the DAISY API: to retrieve project details, datasets for the project, and contracts for the project. If all of these three requests are successful, the DPIA appendix is rendered as a structured set of tables. Again, the rendering of separate tables is done using macros, so if someone needs to change the rendering, they can adjust some of the macros in the appendix template. In the end, one may use just the main macro for retrieving data from DAISY and write the whole rendering part on its own. All macros and parts of the appendix template are documented both in README and using Jinja2 comments.

MarekSuchanek replied:
```suggestion
The macro itself makes three HTTP requests to the DAISY API: to retrieve project details, datasets for the project, and contracts for the project. If all of these three requests are successful, the DPIA appendix is rendered as a structured set of tables. Again, the rendering of separate tables is done using macros, so if someone needs to change the rendering, they can adjust some of the macros in the appendix template. In the end, one may use just the main macro for retrieving data from DAISY and write the whole rendering part on its own. All macros and parts of the appendix template are documented both in README and using Jinja2 comments.
```

froggypaule commented on paper.md:
The EeLP is based on the well-known [Moodle](https://moodle.org) learning management system. As such, it allows to improve the content continuously and add different types of activities and resources. We also plan to include a quiz or feedback form for better user engagement in the future.

MarekSuchanek replied:
```suggestion
The EeLP is based on the well-known [Moodle](https://moodle.org) learning management system. As such, it allows to improve the content continuously and add different types of activities and resources. We also plan to include a quiz or feedback form for better user engagement in the future.
```

froggypaule commented on paper.md:
We completed the two tasks that we set initially. The DPIA Knowledge Model in DSW, together with the related [document template](https://github.com/ds-wizard/dpia-template), can be used to create a project in DSW and answer the questions in order to produce the DPIA document. Both of these deliverables will need final polishing, e.g., styling the document template; they can then be published. We plan to publish them on GitHub (the template) as well as using [DSW Registry](https://registry.ds-wizard.org) service so anyone can easily pull it to their own DSW instance. In this way, both can be easily updated and maintained over time, and others will quickly get the latest versions.

MarekSuchanek replied:
```suggestion
We completed the two tasks that we set initially. The DPIA Knowledge Model in DSW, together with the related [document template](https://github.com/ds-wizard/dpia-template), can be used to create a project in DSW and answer the questions in order to produce the DPIA document. Both of these deliverables will need final polishing, e.g., styling the document template; they can then be published. We plan to publish them on GitHub (the template) as well as using [DSW Registry](https://registry.ds-wizard.org) service so anyone can easily pull it to their own DSW instance. In this way, both can be easily updated and maintained over time, and others will quickly get the latest versions.
```

froggypaule commented on paper.md:
The second task brought new features to DSW. The integration question with a widget still needs some adjustments on the backend side of DSW (making it a new type of integration question); however, it is fully operational. The feature enabling HTTP requests from the document templates is finalized, and it may be enhanced in the future if needed. Both of these features were released in version 3.6.0 of DSW. The widget implementation on the DAISY side also became directly part of the (DAISY??) codebase, and [DSW Integration Widget SDK](https://github.com/ds-wizard/dsw-integration-widget-sdk) has been published on [npmjs.com](https://www.npmjs.com/package/@ds-wizard/integration-widget-sdk). Finally, the DPIA appendix template is ready to be (re-)used and further developed as any other DSW document template.

MarekSuchanek replied:
```suggestion
The second task brought new features to DSW. The integration question with a widget still needs some adjustments on the backend side of DSW (making it a new type of integration question); however, it is fully operational. The feature enabling HTTP requests from the document templates is finalized, and it may be enhanced in the future if needed. Both of these features were released in version 3.6.0 of DSW. The widget implementation on the DAISY side also became directly part of the its codebase, and [DSW Integration Widget SDK](https://github.com/ds-wizard/dsw-integration-widget-sdk) has been published on [npmjs.com](https://www.npmjs.com/package/@ds-wizard/integration-widget-sdk). Finally, the DPIA appendix template is ready to be (re-)used and further developed as any other DSW document template.
```

froggypaule reviewed (commented):
Bravo! well done everyone. It is only after this re-reading of me that I understand the nature of the extension to the API fonctionalities discussed.

MarekSuchanek commented:
> Bravo! well done everyone. It is only after this re-reading of me that I understand the nature of the extension to the API fonctionalities discussed.

Thank you, @froggypaule! I turned the suggested changes into GitHub suggestions in comments. I will then merge them (after others also do the reviews).

Commit:
Update paper.md

Small additions

MarekSuchanek commented on paper.md:
```suggestion
This is obviously not practical and one must thus overcome the need for authentication. To overcome these two issues, we designed a new type of integration question described in the following subsection.
```

MarekSuchanek committed:
Apply suggestions by Paulette

MarekSuchanek committed:
Merge branch 'master' into pinarpink-patch-1

MarekSuchanek committed:
Merge pull request #6 from ds-wizard/pinarpink-patch-1

Update paper.md

MarekSuchanek committed:
Fix consistency and prepositions

There were some inconsistencies (UK/US, curly/straight, etc.Ã)
and minor mistakes (prepositions, articles, etc.) - use Grammarly.

vildead commented on paper.md:
```diff
@@ -38,89 +38,143 @@ affiliations:
 # Abstract
 
-This report summarizes our activities and achievements in connecting Data Stewardship Wizard (DSW) and DAISY during the ELIXIR BioHackathon Europe 2021. As a data information system for GDPR compliance, DAISY is very focused on a single goal – gathering all information required for GDPR compliance. On the other hand, DSW is very flexible and can be used even beyond data management planning. We wo ...
+This report summarises our activities and achievements in integrating the Data Stewardship Wizard (DSW) and Data Information System (DAISY) tools during the ELIXIR BioHackathon Europe 2021. As a data information system for GDPR compliance, DAISY is focused on a single goal – gathering all information required for GDPR accountability of biomedical research projects. On the other hand, DSW is very  ...
```
Just a small comment:
```
results were **polished**
```
 I find this term a bit geeky. Its ok if you want to keep it. Alternatively, it could be replaced by **finalized** or **refined**

MarekSuchanek replied:
Thanks!

```suggestion
This report summarises our activities and achievements in integrating the Data Stewardship Wizard (DSW) and Data Information System (DAISY) tools during the ELIXIR BioHackathon Europe 2021. As a data information system for GDPR compliance, DAISY is focused on a single goal – gathering all information required for GDPR accountability of biomedical research projects. On the other hand, DSW is very flexible and can be used beyond data management planning. We worked on the integration between both tools on two fronts. Firstly, we created a new Knowledge Model in DSW together with a document output template to be able to generate a data protection impact assessment (DPIA). Secondly, we introduced a new integration type between projects in DSW and DAISY that allows the querying of DAISY data upon document generation in DSW. Both of these independent activities brought successful results that were refined and published after the actual BioHackathon. Finally, we provide the related materials as an on-demand training course in the ELIXIR eLearning Platform.
```

vildead reviewed (approved):
I like it!
Thanks a lot ;)

janslifka committed:
Update DSW Daisy question figure

janslifka reviewed (approved):
Nicely written, good job everyone! 👍

barrydjenaba reviewed (approved):
Thanks to everyone, good work.

MarekSuchanek committed:
Apply suggestion from Vilem

Co-authored-by: Vilem Ded <51602017+vildead@users.noreply.github.com>

MarekSuchanek commented:
Thank you all! 👍 I will merge and publish it on BioHackrXiv later today 🚢
